pkg/download/segment/download.go

```diff
@@ -0,0 +1,93 @@
+/*
+ * @license
+ * Copyright 2024 Dynatrace LLC
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package segment
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/dynatrace/dynatrace-configuration-as-code-core/clients/openpipeline"
+	"github.com/dynatrace/dynatrace-configuration-as-code-core/clients/segments"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/log"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/log/field"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/config"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/config/coordinate"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/config/template"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/download/internal/templatetools"
+	project "github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/project/v2"
+)
+
+type Client interface {
```

Comment:
I would call it something more descriptive then Client, as we will split the interface into multiple interfaces.

Reply:
The same discussion I had with @Laubi . This just follows the (bad) design pattern of another endpoint.

Reply:
Is `SegmentClient` better?
I am out of inspiration :)

Reply:
I would go with `DownloadSegmentClient` as there will be also a `deploy.go` and if we call it just `SegmentClient` how do we differentiate?

Reply:
DownloadSegmentClient sounds good to me.

My toughts about naming in general:
If a name needs to be unique in the application (course we expected to use it widely), we should stick with long, descriptive names. 

But this interface shouldn't be used so far from here. And when it is called, it;s going t be with it's package name - `segment.Client`. This 100% complain to the [guidance](https://google.github.io/styleguide/go/guide#naming) in opposite to `segment.DownloadSegmentClient`. `segment.SegmentClient` is even worse.
More probably with our code base that we are going to have package name clashes (e.g. `download/segment.Client`, `deploy/segment.Client`, `delete/segment.Client`).

More on this:
https://go.dev/doc/effective_go#names
https://testing.googleblog.com/2017/10/code-health-identifiernamingpostforworl.html
https://google.github.io/styleguide/go/decisions.html#naming
https://dave.cheney.net/practical-go/presentations/qcon-china.html#_identifiers

Reply:
I would disagree. The interface that is defined here is just an input promise. From the name `segment.Client`, this is unclear. 

Since, as you say, over the project, we have `download/segment.Client`, `deploy/segment.Client`, `delete/segment.Client`, how can we differentiate between them? I would say interfaces should be small and specific, and we should follow SOLID principles -> [Interface segregation](https://en.wikipedia.org/wiki/Interface_segregation_principle), "no code should be forced to depend on methods it does not use". 

That's why I think creating a `segment.Client` is a suboptimal decision. I think it's worth it that I have a long interface name over a wrong design or usage of the interface.

Reply:
What adds to the problem is our complete misuse of the clients within the `ClientSet.` We should not need to have the interface `client.SegmentsClient` at all but use the concrete classes directly. 
With that, the `clientSet` would keep a reference and return the concrete class required, and the download (and deploy and delete) would take very specialized input-interfaces. As far as I understand Go, this should be the way to Go.

Regarding the misuse, this would be a refactoring that is outside of the scope of this PR, but I agree that the naming should be quite specialized.